New function createAsset

Jmr3366 opened the pull request:
### Description of the Change
- Resolves #483 
https://discord.com/channels/296230822262865920/296658523725496320/578194944871759909

New trusted macro function `createAsset(name,imageData)` that will return an `asset:\\< asset ID >` value on success.
Checks are made for data types and only `webp`,`jpg`,`png` are permitted data types.

### Possible Drawbacks
- `createAsset` can allow scriptable asset creation and if implemented improperly, _could be_ campaign breaking.
With great power comes great responsibility.

### Release Notes
Examples: 
`[r: createAsset("Test", < base64 encoded image data >)]`
`[r: createAsset("Test", "https://website.com/image.webp")]`
outputs 
`asset:\\< asset ID >`

(I renamed my branch as I almost overwrote it)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RPTools/maptool/4923)
<!-- Reviewable:end -->

Jmr3366 committed:
Feature #483
https://discord.com/channels/296230822262865920/296658523725496320/578194944871759909

rest functions use macro.return instead of ParserException.

Jmr3366 committed:
Update TokenImage.java

Jmr3366 committed:
Merge branch 'RPTools:develop' into j_develop_new

cwisniew commented:
This needs to ensure the asset is valid before adding it to the campaign

Jmr3366 commented:
I do check file headers for the 3 file formats this function supports.  If there are other checks I can do or add, I will try to comply with those.

cwisniew commented on src/main/java/net/rptools/maptool/client/functions/TokenImage.java:
```diff
@@ -171,6 +173,34 @@ public Object childEvaluate(
+          byte[] imageBytes = Base64.decode(imageString);
+          String imageCheck = new String(imageBytes, 0, 4);
+          /* header check for: webp || jpg || png */
+          if (imageCheck.equals("RIFF") || imageCheck.equals("ÿØÿà") || imageCheck.equals("‰PNG")) {
```
If you make these constants so its more meaningful when people read the code I will approve it :)

Jmr3366 replied:
Is this acceptable or _inline_ with your thoughts?
```
public static final String FILE_HEADER_WEBP = "RIFF";
public static final String FILE_HEADER_JPG = "ÿØÿà";
public static final String FILE_HEADER_PNG = "‰PNG";
  ....
if (imageCheck.equals(FILE_HEADER_WEBP) || imageCheck.equals(FILE_HEADER_JPG) || imageCheck.equals(FILE_HEADER_PNG)) {```

edit - double posted . .  .

Jmr3366 replied:
Is this in-line with what you're wanting or am I _off_ :) ?
``` if (imageCheck.equals(FILE_HEADER_WEBP) || imageCheck.equals(FILE_HEADER_JPG) || imageCheck.equals(FILE_HEADER_PNG)) {```

Jmr3366 replied:
spotless, I hate thee . . .
Added the constants, better checks re: empty string and npe catch
Added "https:" support for the same file types.

Jmr3366 committed:
Merge branch 'RPTools:develop' into j_develop_createAsset

Jmr3366 committed:
Update TokenImage.java

Added request changes to use constants
Changed empty variable check
Added NPE check for bad bytes

Jmr3366 committed:
Update TokenImage.java

spotless proper

Jmr3366 committed:
Resolves #483

Jmr3366 committed:
Merge branch 'develop' into j_develop_createAsset

Jmr3366 committed:
Merge remote-tracking branch 'origin/j_develop_createAsset' into j_develop_createAsset

# Conflicts:
#	src/main/java/net/rptools/maptool/client/functions/TokenImage.java

Jmr3366 committed:
Update TokenImage.java

spotless I hate thee

Jmr3366 committed:
Merge remote-tracking branch 'origin/j_develop_createAsset' into j_develop_createAsset

Jmr3366 committed:
Resolves #483

Jmr3366 committed:
Resolves #483

Jmr3366 committed:
Resolves #483 added https url as an option
